### PR TITLE
Builder pattern realized with tests

### PR DESCRIPTION
**Pull request by martinmilev**
Ready for review

### CODE REVIEW

**Comment by martinmilev** [comment]
Ready for review.

File: src/com/clouway/builder/Order.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  public List<OrderItem> getItems() {
+    return items;
+  }
+
+
+  public boolean ontime() {
+    return orderDate.after(orderDeliveryDate);
+  }
+
+  static final class OrderBuilder {
```
Comment (freekman) [inline src/com/clouway/builder/Order.java]: can you tell me why is the OrderBuilder inside the Order

File: src/com/clouway/builder/OrderItem.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public final class OrderItem {
+  private String productName;
+  private String measureUnit;
+  private Double quantity;
+  private Double price;
+
+  private OrderItem(ItemBuilder builder) {
```
Comment (freekman) [inline src/com/clouway/builder/OrderItem.java]: please explain to me the purpose of this procedure

Comment (martinmilev) [inline src/com/clouway/builder/OrderItem.java]: That way you can create new object only from the builder.

File: src/com/clouway/builder/Order.java (hunk 1 above)
Comment (martinmilev) [inline src/com/clouway/builder/Order.java]: That way only the Builder can be instantiated.

File: src/com/clouway/builder/OrderItem.java (hunk 2 above)
Comment (martinmilev) [inline src/com/clouway/builder/OrderItem.java]: The constructor is private and takes an instance of the Builder. 
The fields are initialized to the values from the Builder.

File: src/com/clouway/builder/ItemBuilder.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
+package com.clouway.builder;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public class ItemBuilder {
+  private String productName;
```
Comment (freekman) [inline src/com/clouway/builder/ItemBuilder.java]: for the builder it would be nice to add default value for the fields

Comment (martinmilev) [inline src/com/clouway/builder/ItemBuilder.java]: Fixed.

File: src/com/clouway/builder/Order.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+public final class Order {
+  private Long orderId;
+  private String customerName;
+  private String customerAddress;
+  private Date orderDate;
+  private Date orderCreationDate;
+  private Date orderDeliveryDate;
+  private List<OrderItem> items;
+
+  public Order(OrderBuilder builder) {
```
Comment (freekman) [inline src/com/clouway/builder/Order.java]: why is the order still getting the ```OrderBuilder``` in the constructor

File: src/com/clouway/builder/OrderItem.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public final class OrderItem {
+  private String productName;
+  private String measureUnit;
+  private Double quantity;
+  private Double price;
+
+  public OrderItem(ItemBuilder builder) {
```
Comment (freekman) [inline src/com/clouway/builder/OrderItem.java]: .

**Pull request closed without merging**